### PR TITLE
Add communications bandwidth to benchmarks

### PR DESCRIPTION
**Pull request by pentschev**
The existing throughput would only give an overview of total data processed for the complete workflow, but no insight on communications which the new bandwidth value now does. Additionally, moved common peer-to-peer bandwidth computation as utility function.

Depends on #937

### CODE REVIEW

File: dask_cuda/benchmarks/local_cupy.py [hunk 1]
```diff
@@ -210,63 +212,72 @@ async def run(args):
...
+                times.append(took)
+                print_key_value(key=f"{m}", value=npartitions)
+            times = np.asarray(times)
+            bandwidths_all = np.asarray(bandwidths_all)
+            print_separator(separator="=")
+            print_key_value(
+                key="Bandwidth",
+                value=f"{format_bytes(bandwidths_all.mean())}/s +/- "
+                f"{format_bytes(bandwidths_all.std())}/s",
+            )
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cupy.py:264]: FWIW, and this is just the same as before, the arithmetic mean is generally the wrong thing for rate-like data, and the more correct option is the harmonic mean `len(xs)/sum(1/xs)`.

File: dask_cuda/benchmarks/local_cudf_shuffle.py [hunk 2]
```diff
@@ -171,12 +156,34 @@ def main(args):
...
+    times = np.asarray(times)
+    bandwidths_all = np.asarray(bandwidths_all)
     print_separator(separator="=")
+    print_key_value(
+        key="Throughput (Total)",
+        value=f"{format_bytes(t_p.mean())} +/- {format_bytes(t_p.std())}",
+    )
+    print_key_value(
+        key="Throughput (Comms only)",
+        value=f"{format_bytes(bandwidths_all.mean())} +/- "
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cudf_shuffle.py]: Whereas this bandwidth measure is the mean of the all the p2p data movement observeed by dask?

File: dask_cuda/benchmarks/local_cudf_shuffle.py [hunk 3]
```diff
@@ -171,12 +156,34 @@ def main(args):
...
+        t_p.append(throughput)
         print_key_value(key=f"{m}", value=f"{format_bytes(throughput)}/s")
         t_runs[idx] = float(format_bytes(throughput).split(" ")[0])
+    t_p = np.asarray(t_p)
+    times = np.asarray(times)
+    bandwidths_all = np.asarray(bandwidths_all)
     print_separator(separator="=")
+    print_key_value(
+        key="Throughput (Total)",
+        value=f"{format_bytes(t_p.mean())} +/- {format_bytes(t_p.std())}",
```
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/benchmarks/local_cudf_shuffle.py]: Perhaps it is worth clarifying these differences?

**Review by wence- — COMMENTED** [review]
Minor queries

File: dask_cuda/benchmarks/local_cupy.py (hunk 1 above)
Comment (pentschev) [reply dask_cuda/benchmarks/local_cupy.py:266]: Yeah, I was actually wondering if that was correct but didn't spend the time in investigating what would be the correct way to compute. Harmonic means do make sense indeed. However, it looks depressing 🙁 

```
Bandwidth                 | 10.18 GiB/s +/- 8.16 GiB/s
BandwidthH                | 3.47 GiB/s +/- 1.25 GiB/s
```

We _seemingly_ go down from 10.18GB/s to 3.47GB/s, of course the former is incorrect anyway, but it is sad.

I now replaced mean and standard deviation in https://github.com/rapidsai/dask-cuda/pull/938/commits/268f893b98faaccc87bea4f97060d502d513a15a, I am not sure whether standard deviation is correct, can you double-check?

File: dask_cuda/benchmarks/local_cudf_shuffle.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply dask_cuda/benchmarks/local_cudf_shuffle.py]: That is correct, although you commented on an old commit, the newest modifications read "Throughput" and "Bandwidth", do you think that naming is still unclear?

File: dask_cuda/benchmarks/local_cudf_shuffle.py (hunk 2 above)
Comment (pentschev) [reply dask_cuda/benchmarks/local_cudf_shuffle.py]: Correct.

File: dask_cuda/benchmarks/local_cudf_merge.py [hunk 4]
```diff
@@ -308,12 +293,18 @@ def main(args):
...
     print_key_value(
-        key="Throughput",
-        value=f"{format_bytes(t_p.mean())} +/- {format_bytes(t_p.std()) }",
+        key="Throughput (Total)",
+        value=f"{format_bytes(t_p.mean())} +/- {format_bytes(t_p.std())}",
+    )
+    print_key_value(
+        key="Throughput (Comms only)",
+        value=f"{format_bytes(bandwidths_all.mean())} +/- "
+        f"{format_bytes(bandwidths_all.std())}",
```
[the remark this answers is not in the document]
Comment (pentschev) [reply dask_cuda/benchmarks/local_cudf_merge.py]: Doesn't seem to be the case, although it's possible this is only for normal message passing (e.g., heartbeats):

```
$ DASK_DISTRIBUTED__COMM__COMPRESSION=None OPENBLAS_NUM_THREADS=1 python dask_cuda/benchmarks/local_cudf_merge.py -p ucx -d 0,1,2,3,4,5,6,7 -c 1_000_000 --runs 30 -b explicit-comms
...
Throughput                | 1.07 GiB/s +/- 209.63 MiB/s
Bandwidth                 | 103.27 MiB/s +/- 90.80 MiB/s
```

File: dask_cuda/benchmarks/local_cupy.py (hunk 1 above)
Comment (wence-) [reply dask_cuda/benchmarks/local_cupy.py:266]: Standard deviations of inverse random variables is a huge can of worms :(, but as per [this stats.stackexchange post](https://stats.stackexchange.com/q/7471) there are a few options. Suppose we have our rates $X_i, \quad i=[1,\dots,n]$, then define the reciprocal weights $Y_i$ and calculate

$$
\begin{align}
\mu_Y &= \frac{1}{n}\sum_i Y_i \\
\text{Var}(Y) &= \frac{1}{n}\sum_i (Y_i - \mu_Y)^2
\end{align}
$$

Then the variance of the harmonic mean of $X_i$ ($H_X := \frac{1}{\mu_Y}$) is 

$$
\text{Var}_H = \frac{\text{Var}(Y)}{n \mu_Y^4}
$$

The above-linked post shows that this is a good estimate for uniformly distributed data at least.

So I think we can calculate harmonic mean and standard deviation as

```python
def hmean(data):
    return 1/np.mean(1/data)

def hstd(data):
    rmean = np.mean(1/data)
    rvar = np.var(1/data)
    return np.sqrt(rvar / (len(data) * rmean**4))
```

File: dask_cuda/benchmarks/local_cudf_shuffle.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/benchmarks/local_cudf_shuffle.py]: I think those are OK.

File: dask_cuda/benchmarks/local_cudf_merge.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/benchmarks/local_cudf_merge.py]: So this must be background messaging since I don't think that the bandwidth can be lower than the throughput if we're actually measuring data movement of the computation.

File: dask_cuda/benchmarks/local_cupy.py (hunk 1 above)
Comment (pentschev) [reply dask_cuda/benchmarks/local_cupy.py:266]: Thanks, I saw that yesterday but wasn't totally sure that would be necessarily correct either. I've made the changes in https://github.com/rapidsai/dask-cuda/pull/938/commits/d120cd3e042382553b1b187a77abb6b08fe12c41 , thanks for the code as well!

File: dask_cuda/benchmarks/local_cudf_merge.py [hunk 5]
```diff
@@ -308,12 +295,18 @@ def main(args):
...
-    t_p = numpy.asarray(t_p)
-    times = numpy.asarray(times)
+    t_p = np.asarray(t_p)
+    times = np.asarray(times)
+    bandwidths_all = np.asarray(bandwidths_all)
     print_separator(separator="=")
     print_key_value(
         key="Throughput",
-        value=f"{format_bytes(t_p.mean())} +/- {format_bytes(t_p.std()) }",
+        value=f"{format_bytes(hmean(t_p))}/s +/- {format_bytes(hstd(t_p))}/s",
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cudf_merge.py:304]: I had a check to compare with how standard errors are propagated for reciprocals and these formulae are consistent with that approach. For this case `data_processed` is the same for every run and the times vary. So we can compute the mean throughput with `data_processed / np.mean(times)` and the standard deviation with `data_processed * np.std(times) / np.mean(times)**2`. But these are just special cases of the harmonic approach, so I'm happy.

File: dask_cuda/benchmarks/local_cupy.py [hunk 6]
```diff
@@ -263,11 +249,25 @@ async def run(args):
...
             print_key_value(key="Worker thread(s)", value=f"{args.threads_per_worker}")
             print_separator(separator="=")
-            print_key_value(key="Wall clock", value="npartitions")
+            print_key_value(key="Wall clock", value="Partitions")
             print_separator(separator="-")
-            for (took, npartitions) in took_list:
-                t = format_time(took)
-                print_key_value(key=f"{t}", value=f"{npartitions}")
+            times = []
+            for idx, (took, npartitions) in enumerate(took_list):
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cupy.py:255]: `idx` unused.

File: dask_cuda/benchmarks/local_cupy_map_overlap.py [hunk 7]
```diff
@@ -148,11 +134,25 @@ async def run(args):
...
             print_key_value(key="Worker thread(s)", value=f"{args.threads_per_worker}")
             print_separator(separator="=")
-            print_key_value(key="Wall clock", value="npartitions")
+            print_key_value(key="Wall clock", value="Partitions")
             print_separator(separator="-")
-            for (took, npartitions) in took_list:
-                t = format_time(took)
-                print_key_value(key=f"{t}", value=f"{npartitions}")
+            times = []
+            for idx, (took, npartitions) in enumerate(took_list):
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cupy_map_overlap.py:140]: `idx` unused

File: dask_cuda/benchmarks/local_cupy.py [hunk 8]
```diff
@@ -263,11 +249,25 @@ async def run(args):
...
+                m = format_time(took)
+                times.append(took)
+                print_key_value(key=f"{m}", value=npartitions)
+            times = np.asarray(times)
+            bandwidths_all = np.asarray(bandwidths_all)
+            print_separator(separator="=")
+            print_key_value(
```
Comment (wence-) [inline dask_cuda/benchmarks/local_cupy.py:262]: I think we can deduce throughput for these operations as well, but it would need to be done for each op in turn, and we'd need to pick a cache model. For example does `(A + A.T).sum()` move `A.nbytes` of data or `2*A.nbytes` (or more)?

**Review by wence- — APPROVED** [review]
nitpicks, but otherwise fine

**Review by quasiben — APPROVED** [review]
based on @wence- 

Probably good to address Lawrence's comments before merging in

File: dask_cuda/benchmarks/local_cupy.py (hunk 8 above)
Comment (pentschev) [reply dask_cuda/benchmarks/local_cupy.py:262]: The amount of data that gets moved (if by move you mean the same as transferring between workers) in the cuDF benchmarks is irrelevant for throughput, instead the throughput is based on the `data_processed` which is a sum of the DataFrames sizes:

https://github.com/rapidsai/dask-cuda/blob/a2e322362515bf612f9368eee99e150e6fa49437/dask_cuda/benchmarks/local_cudf_merge.py#L169-L170

In the same manner, I believe the equivalent `data_processed` for CuPy benchmarks here would be the sum of sizes of input arrays, e.g., `x.nbytes` for 

https://github.com/rapidsai/dask-cuda/blob/a2e322362515bf612f9368eee99e150e6fa49437/dask_cuda/benchmarks/local_cupy.py#L35

and `x.nbytes + y.nbytes` for

https://github.com/rapidsai/dask-cuda/blob/a2e322362515bf612f9368eee99e150e6fa49437/dask_cuda/benchmarks/local_cupy.py#L44-L45

IOW, the operation `A.T` size is not an input and thus doesn't get accounted for in `data_processed`. Does the above make sense?

Comment (wence-) [reply dask_cuda/benchmarks/local_cupy.py:262]: I meant data movement in the local processing, but I guess this is a slightly philosophical argument. I think I was thinking from the point of view of what is the streaming data cost? e.g. if I just compute `B = A.T` then I need to stream `A` and stream the output to `B` so the data movement is 2*A.nbytes (on cpus at least it would normally be 3*A.nbytes since caches are usually write-allocate without hinting that the output can be done with a non-temporal store). This kind of metric tells us something about the performance on the hardware.

But I agree that a throughput of "this is how much data I started with, how fast can I process it?" is also a good metric, and is generally the one we actually care about since that's the wallclock time question. 

In conclusion: happy to go with sum of input array sizes.

File: dask_cuda/benchmarks/local_cupy_map_overlap.py (hunk 7 above)
Comment (pentschev) [reply dask_cuda/benchmarks/local_cupy_map_overlap.py]: Done in https://github.com/rapidsai/dask-cuda/commit/dc1a58715b0053eac7662229e8b892707a7d82e3

**Comment by wence-** [comment]
> @wence- I think all your suggestions/requests have been addressed, do you think we're good to merge?

Yes, go ahead.

**Comment by pentschev** [comment]
Thanks @wence- and @quasiben for reviews!
